type-auto-algorithm.md

```diff
@@ -0,0 +1,57 @@
+# `--type=auto` Module Type Detection Algorithm
+
+This is the algorithm, in psuedocode, for the `--type=auto`/`-a` Node startup option described [here](./README.md). It is based on detailed syntax rules outlined [here](https://github.com/nodejs/node-eps/issues/57#issuecomment-300870976).
+
+This algorithm relies on the fact that it is evaluating an initial entry point—the very first file loaded for a program. Affirmatively detecting CommonJS would be virtually impossible otherwise, as it would no longer be a safe assumption that a reference to a globally defined `require` confirms that a file is CommonJS (since a file loaded earlier in the program could have defined `require` in the global scope). Additionally, many files (either CommonJS or ESM) that are not entry points are often ambiguous, for example containing only `console.log` statements or references to user-defined globals. For this algorithm to be both confident (no false positives) and inclusive (few files judged ambiguous) it must be restricted to initial entry points.
+
+## Step 1: Parse
+
+The initial entry point file must be parsed so that its syntax may be analyzed.
+
+## Step 2: Look for signifiers of ESM JavaScript
+
+If the source code contains either of the following:
+
+- An `import` statement, like `import "x";`
+
+- An `export` statement, like `export {};`
+
+Then the file is unambiguously ESM. Stop and execute it as ESM JavaScript.
+
+These are the only two syntactical elements that are unambiguously ESM. The `import` expression, `import("x")`, is allowed in CommonJS.
+
+If the source code did not contain an ESM signifier, continue:
+
+## Step 3: Look for signifiers of CommonJS JavaScript
+
+If the source code contains any of the following:
+
+- A reference to a variable named `__filename` without a declaration for that variable.
+
+- A reference to a variable named `__dirname` without a declaration for that variable.
+
+- A reference to a variable named `require` without a declaration for that variable.
+
+- A reference to a variable named `module` without a declaration for that variable.
```

Comment:
this could be a global used in ESM.

Reply:
it would make sense to look for well known names being accessed off. member expressions like module.exports or module.parent

Reply:
Wouldn't checking for a reference to `module` include those?

Reply:
@geoffreybooth sure, I'm just saying I think it makes sense to go as deep as possible. checking for module alone is fine (assuming entrypoint), but more data is always good

Reply:
In this case, the more things we check for the slower Node starts up, so I think we want to keep it limited. Checking for accessors off `module` would be redundant.

Reply:
`global.module = {}` now all ESM files can access `module`.

No non-keyword identifier can *possibly* be used to disambiguate reliably.